course-content/21_getting_data_with_pandas.qmd

```diff
@@ -0,0 +1,427 @@
+
+First, we need to make sure that we import the library into Python so we can use it:
+
+```{python}
+import pandas as pd
+```
+
+Objects in pandas are typically two-dimensional tables called **dataframes**. You can think of this as being similar to a single spreadsheet in excel. Each column (called **series**) in the dataframe can have a name, as well as an indexing value. 
+
+<center>
+    <img src="images/getting-data-with-pandas/dataframe.png" 
+    width=600 style="margin:auto"/>
+    <p style="text-align: center">
+        A pandas dataframe
+    </p>
+</center>
+
+We can create both single series in pandas or full dataframes. Let's consider a single series first. We can create this using the pandas `Series` method, as well as a list of elements we wish to include in our series. 
+
+```{python}
+series_x = pd.Series([1, 2, 3])
+series_x
+```
+
+When this displays in python, we see 2 columns. The left column is always the index, while the right column contains the data values. The series is also given a `dtype`. This is similar to the datatypes we considered previously, and tells us what type of data is held inside the series. When a series is first created, pandas will try to guess the `dtype` for the series based on the individual elements inside it. We can make a series made of `float` numbers:
+
+```{python}
+series_y = pd.Series([1, 2.0, 3.14])
+series_y
+```
+
+Notice that here, the first element was actually converted from an `int` to a `float` by the program! This will often happen when working with `int` and `float` type objects. 
+
+Next, a series of `str` elements will default to the `dtype` of `object`.
+
+
+```{python}
+series_z = pd.Series(['settlers', 'of', 'catan'])
+series_z
+```
+
+The `dtype: object` is pandas 'catchall' Series type, and we want to be cautious when we see this! It is also the Series dtype used when we have mixed data:
+
+
+```{python}
+series_mixed = pd.Series([1, 1.0, 'one'])
+series_mixed
+```
+
+When pandas sees data that cannot be obviously converted into a single datatype, it leaves each individual entry alone, as whatever its original datatype happened to be. Be careful when you see this! Many operations on pandas dataframes apply to the entire column (similar to how a new excel column is often created with a function that applies to a previous column). If this operation is built to only work on a single datatype, we might run into errors! To avoid this, we can utilize the `.astype()` methods. Possible arguments for `astype()` include the standard datatypes:
+
+ * `int`
+ * `float`
+ * `str`
+ * `bool`
+ 
+However if you need to be more specific, you can be as well. These arguments are all required to be included in quotations, as they refer to aliases for pandas specific datatypes:
+
+ * `'Int8'`, `'Int32'`, `'Int64'` provide nullable integer types (note the capital I!)
```

Comment:
maybe explain what nullable means

Reply:
Good call, I've added more to the tool tip box on NaN just below this bit.